core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -228,7 +249,9 @@ public static List<Pair<Integer, ValueReader<?>>> buildReadPlan(
       if (constant != null) {
         readPlan.add(Pair.of(pos, ValueReaders.constant(constant)));
       } else if (field.initialDefault() != null) {
-        readPlan.add(Pair.of(pos, ValueReaders.constant(field.initialDefault())));
+        readPlan.add(
+            Pair.of(
+                pos, ValueReaders.constant(convert.apply(field.type(), field.initialDefault()))));
```

Comment:
This raises a question: should we also convert the constants passed by `idToConstant`?

I think we _should_ but not yet. The problem is that this conversion probably can't be done twice because of assumptions in the conversion (like casting strings to `UTF8String` in Spark) so it is worth a separate PR to refactor all of the conversions to be done in one place. That will also be different for each file format so it could get messy.

Reply:
Let's do that in a separate PR indeed 👍